### ISSUE TITLE
[feat.]: Methods to delete the registered modules

### ISSUE THREAD

**Issue by chenzhuofu** [WasmEdge/WasmEdge#2481]
## Motivation

In WasmEdge C-SDK, there is no provisioned method to remove the modules registerd by VM.
 
I can understand the reason is that it's vm owns those modules. However it's still such a wasting behavior if u don't delete the registered module after using it.

Maybe we can create some methods like this:
```cpp
WASMEDGE_CAPI_EXPORT extern void
WasmEdge_VMDeleteRegisteredModule(const WasmEdge_VMContext *Cxt,
                               const WasmEdge_String ModuleName);
```

**Comment by aetiurf** [comment]
And when we invoke this function, the VM could release its ownership and delete this named module.

**Comment by hydai** [comment]
@q82419 
What do you think?

**Comment by q82419** [comment]
As discussed before, the module instances may have dependencies: module A imports the functions of module B when instantiating.
Therefore, this function is not safe if delete the specific registered module, or it may cause core dump when execution.
As the steps in [module instantiation](https://webassembly.github.io/spec/core/exec/modules.html#instantiation), the [external value (the instance address in store)](https://webassembly.github.io/spec/core/exec/runtime.html#syntax-externval) should be able to be found. Therefore in WasmEdge we did the strict linking to reduce the instance searching time when executing.
I think this function is only available if we implement the dynamic WASM module instance linker.

You can use the `WasmEdge_VMCleanUp` to delete all registered modules without WASI and plug-ins.

**Comment by aetiurf** [comment]
> module A imports the functions of module B when instantiating

Maybe I could call it "synchronization problem" ?

IMO, this problem should be considered by developers. That is, When a developer calls `WasmEdge_VMDeleteRegisteredModule`, he should ensure that the registered module isn't or will never become the dependency of other modules (from other threads) anymore.

**Comment by q82419** [comment]
> > module A imports the functions of module B when instantiating
> 
> Maybe I could call it "synchronization problem" ?
> 
> IMO, this problem should be considered by developers. That is, When a developer calls `WasmEdge_VMDeleteRegisteredModule`, he should ensure that the registered module isn't or will never become the dependency of other modules (from other threads) anymore.
> 
> It's just like the ModuleInstance owned by developer: one can call `WasmEdge_ModuleInstanceDelete` whenver he want.

Yes, you're right, developers should ensure the life-cycle of the modules.
That's why we have `WasmEdge_ModuleInstanceDelete` API -- developers can delete the module if they use the `WasmEdge_ExecutorContext` but not the `WasmEdge_VMContext` APIs to run WASM.

In the design of WasmEdge, not like the other WASM runtimes, we provide a more convenient `VM` objects to let developers to instantiate and run WASM more easily and quickly than the traditional WASM C API. In this case, I think the `VM` should give more protections to prevent from segmentation fault, such as this module deletion case.
Of course if you use the `Executor` and `Store` to instantiate a WASM module, you can easily to this detailed control.
I think we can implement this API if we implemented the dependency tree of registered modules.

cc @hydai  do you think we can implement this API directly or after implementing the dependency tree of modules in store?

**Comment by aetiurf** [comment]
> > > module A imports the functions of module B when instantiating
> > 
> > 
> > Maybe I could call it "synchronization problem" ?
> > IMO, this problem should be considered by developers. That is, When a developer calls `WasmEdge_VMDeleteRegisteredModule`, he should ensure that the registered module isn't or will never become the dependency of other modules (from other threads) anymore.
> > It's just like the ModuleInstance owned by developer: one can call `WasmEdge_ModuleInstanceDelete` whenver he want.
> 
> Yes, you're right, developers should ensure the life-cycle of the modules. That's why we have `WasmEdge_ModuleInstanceDelete` API -- developers can delete the module if they use the `WasmEdge_ExecutorContext` but not the `WasmEdge_VMContext` APIs to run WASM.
> 
> In the design of WasmEdge, not like the other WASM runtimes, we provide a more convenient `VM` objects to let developers to instantiate and run WASM more easily and quickly than the traditional WASM C API. In this case, I think the `VM` should give more protections to prevent from segmentation fault, such as this module deletion case. Of course if you use the `Executor` and `Store` to instantiate a WASM module, you can easily to this detailed control. I think we can implement this API if we implemented the dependency tree of registered modules.
> 
> cc @hydai do you think we can implement this API directly or after implementing the dependency tree of modules in store?

@hydai Just in case this issue has escaped your attention, I would like to kindly remind you about it.

**Comment by hydai** [comment]
> do you think we can implement this API directly or after implementing the dependency tree of modules in store?

We can have this API first with a big warning message and then implement the dependency tree in the future. Maybe we can have it in the 0.13 release?

**Comment by hydai** [comment]
Since this is not a critical issue, I mark it as a `good first issue`. A PR is welcome.

**Comment by vishwahaha** [comment]
Hi @hydai, I would like to work on this.
Since I'm new to open source, would you recommend any material to go through before starting to work on this?
I have gone through the documentation and set things up on my machine.

**Comment by hydai** [comment]
Hi @vishwahaha 
Since this one is a C API header and C++ implementation, you will need a basic idea about these languages. And one thing is that you can directly add this function and do some tests on it. Please feel free to ask anything if you have any questions.

**Comment by vishwahaha** [comment]
@hydai 
I have written the function for now, Can you tell me what/how should I include the "big warning" message?
Also should I include the test in the PR later?

**Comment by vishwahaha** [comment]
Hi @hydai 
I was writing tests for `WasmEdge_VMDeleteRegisteredModule`, I noticed that `WasmEdge_VMListRegisteredModuleLength` return `12` after instantiating with `WasmEdge_VMCreate` in the APICoreTest. Can you tell me why this happens? Is this related to the initial configuration object?
Also, should I include the test in the PR?

**Comment by hydai** [comment]
Hi @vishwahaha
Sorry for the late response, I must miss your previous messages.

> @hydai I have written the function for now, Can you tell me what/how should I include the "big warning" message? Also should I include the test in the PR later?

1. You should add a warning message in the comments before the function.
2. Yes, please.



> Hi @hydai I was writing tests for `WasmEdge_VMDeleteRegisteredModule`, I noticed that `WasmEdge_VMListRegisteredModuleLength` return `12` after instantiating with `WasmEdge_VMCreate` in the APICoreTest. Can you tell me why this happens? Is this related to the initial configuration object? Also, should I include the test in the PR?

1. There are some pre-registered modules, such as WASI. It's related to the initial configuration.
2. You should register a new one and delete it. Not delete the existing one.

**Comment by KhawajaFashi** [comment]
Hello @hydai , I would like to work on this.
Since I'm new to open source, would you recommend any material to go through before starting to work on this?
I have gone through the documentation and set things up on my machine.

**Comment by hydai** [comment]
> Hello @hydai , I would like to work on this. Since I'm new to open source, would you recommend any material to go through before starting to work on this? I have gone through the documentation and set things up on my machine.

Assuming you know how C SDK works, if not, checking the C SDK related files and docs first.

**Comment by KhawajaFashi** [comment]
> > Hello @hydai , I would like to work on this. Since I'm new to open source, would you recommend any material to go through before starting to work on this? I have gone through the documentation and set things up on my machine.
> 
> Assuming you know how C SDK works, if not, checking the C SDK related files and docs first.

Yes I have read C SDK docs how to proceed next.

**Comment by ShigrafS** [comment]
Hi! @hydai 
Is this issue still available to work on? 
I would like to work on it.

**Comment by hydai** [comment]
It's still available. Assigned.
